### PR TITLE
Fixup for failing multiapp solves during picard iteration

### LINKED ISSUE #5827: Solution vectors are not restored when solve fails in sub-apps while doing Picard iterations
When running BISON (master) and RELAP-7 (sub-app), it can happen that RELAP-7 gets into a non-physical region which results in a NaN in the solution vector. MOOSE correctly cuts the time step in a half and tries to resolve, however, the initial vector on RELAP-7 side still has a NaN, so the solve aborts immediately, time steps gets cut again and so on until we reach dt_min.

### PR DESCRIPTION
**Pull request by friedmud**
We needed to make sure that we restore multiapps in the case of failure.
Even if that MultiApp would normally not need restoration.

 closes #5827

### CODE REVIEW

**Comment by friedmud** [comment]
I need to devise a more stable way to test this...

File: test/src/kernels/NanAtCountKernel.C [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+Real
+NanAtCountKernel::computeQpResidual()
+{
+  _count++;
+
+  if (_print_count)
+    _console << "NanAtCountKernel " << name() << " count: " << _count << "\n";
+
+  if (_count == _count_to_nan)
+    return std::numeric_limits<Real>::infinity();
```
Comment (permcody) [inline test/src/kernels/NanAtCountKernel.C]: What? No 9999999?!

Comment (friedmud) [inline test/src/kernels/NanAtCountKernel.C]: :-)

**Comment by andrsd** [comment]
:+1:

**Comment by friedmud** [comment]
@andrsd did you try it on the real problem?  I'm interested to know if it works...

**Comment by andrsd** [comment]
@friedmud I am going to...

**Pull request merged**